petastorm/pytorch.py

```diff
@@ -423,3 +367,139 @@ def __enter__(self):
+    """
+
+    def __init__(self, reader, batch_size=1,
+                 transform_fn=None,
+                 num_epochs=1,
+                 seed=0,
+                 rows_capacity=1024,
+                 shuffle=False):
+        """
+        Initializes a data loader object.
+        This class does not support special types that are not supported in PyTorch (decimal/string).
+        :param reader: petastorm Reader instance
+        :param batch_size: the number of items to return per batch; factored into the len() of this reader
+        :param transform_fn: an optional callable to convert batches from the reader to PyTorch tensors
+        :param num_epochs: number of epochs.
+        :param seed: random seed used to shuffle.
+        :param rows_capacity: number of rows to be loaded into memory.
+        :param shuffle: If ``True``, indices will be shuffled in every epoch.
+        """
+        super(InMemBatchedDataLoader, self).__init__()
+        self._reader = reader
+        self._batch_size = batch_size
+        self._transform_fn = transform_fn or torch.as_tensor
+        self._num_epochs = num_epochs
+        self._seed = seed
+        self._rows_capacity = rows_capacity
+        self._shuffle = shuffle
+        self._in_iter = False
+
+    def __iter__(self):
+        """
+        The Data Loader iterator stops the for-loop when num_epochs is reached.
+        """
+        if self._in_iter:
+            raise RuntimeError("InMemBatchedDataLoader couldn't be used multiple times, please\
+                    specify total number of epochs using num_epochs in constructor.")
+        self._in_iter = True
+        try:
+            keys, buffer = _load_rows_into_mem(self._reader, self._transform_fn, self._rows_capacity)
+            for epoch in range(self._num_epochs):
+                size = len(buffer[0])
+                if self._shuffle:
+                    # Deterministically shuffle based on seed and current epoch id.
+                    g = torch.Generator()
+                    g.manual_seed(self._seed + epoch)
+                    indices = torch.randperm(size, generator=g).tolist()
+                else:
+                    indices = list(range(size))
+
+                # Sample batches
+                for i in range(0, len(indices), self._batch_size):
+                    idx = indices[i:i+self._batch_size]
+                    batch = [v[idx] for v in buffer]
+                    size -= len(batch[0])
+                    batch = dict(zip(keys, batch))
+                    yield batch
+                assert size == 0
+        except Exception as e:
+            logger.error('Iteration on InMemBatchedDataLoader raise error: %s', repr(e))
+            raise
+
+    # Functions needed to treat data loader as a context manager
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._reader.join()
```

Comment:
Would we get into a deadlock if a user instantiated `InMemBatchedDataLoader` but never iterated? Perhaps should unit-test that case.

Reply:
- reader is loaded in constructor
- reader is stopped after expected rows are loaded
- work_pool is joined in reader.stop(), here we can leave it nothing just `pass`.
Maybe this is more straightforward?

Reply:
I added a timeout test to check possible deadlock.